### PR TITLE
Implement Win32 threading functions

### LINKED ISSUE #4: Windows threads support
Would be really nice to see win threads support here! I might do it because I'm quite desperate looking for C-only cross-platform threading lib.

### PR DESCRIPTION
**Pull request by oold**
This is my implementation for Win32. I've only tested this on my system for x86 and x64 with MSVC and clang-cl, but it's working so far.
As part of the implementation, I have written some new tests and improved the error reporting.

There are new global initialization and deinitialization functions for Win32 and some variables. Precisely, we need a list for the TSS dtors and another list holding the thread handles to prevent the system from reaping the dead threads until `thrd_join()` or `thrd_detach()` are called. I've chosen a relatively simple approach here, using critical sections, and two linked lists.

"Thunks" were necessary to make `thrd_create()` and `call_once()` work. `thrd_create()` needs a `__stdcall` function as well as `call_once()`. In addition, the callback for `call_once()` has a vastly different prototype than the one for `InitOnceExecuteOnce()`. In the case of `thrd_create()`, we also have to run the TSS dtors on exit, but that wouldn't happen if the thread function `return`s, so there's a call to the cleanup function inside the thunk.

Resolves #4.

### CODE REVIEW

**Comment by jtsiomb** [comment]
Thank you! Very interesting.

The reason I initially decided to make this a header-only thing with static inline functions, was that C11 threads are mostly 1-1 with POSIX threads, and the implementation of every function is pretty much an one-liner. This is clearly not the case with win32, which makes the header file go from 280 lines to 1000, and now with the extra complication of globals which need to be defined once. This is no longer simple enough for a single-header wrapper. Still, win32 support is useful. 

To solve both issues at once, I propose we refactor your win32 implementation to go in its own C source file. UNIX users can still simply drop just the header in their projects, while windows users would also need to add the c11threads_win32.c file to their build. To facilitate that, the header should put all the static inline functions in the `#else` part of a `#if defined(_WIN32) && !defined(C11THREADS_PTHREAD_WIN32)` conditional, with function declarations only in the main part. The reason for the extra `C11THREADS_PTHREAD_WIN32` define is to allow the option of using pthread-win32 on windows instead of the win32 API code.

A second change I'd like to do is to move all the MSVC trash in a subdirectory to keep them out of sight, if at all possible.

Are you willing to make these changes to your pull request, or would you prefer I just merge this in a branch, and do them myself as time allows?

Edit: also ideally at some point I'd like to drop the vista requirement, but we can leave that for a future change.

**Comment by oold** [comment]
I'll try to work on those changes in this pull request. It's definitely doable for me and the Vista requirement can also be lowered to XP if I work around the `call_once()`-related stuff. Might also be worth considering a callable initialization function for the globals.

**Comment by oold** [comment]
> **Windows Server 2003 and Windows XP:** Condition variables are not supported.

Oh. Well, that's unfortunate.

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h:27]: Is it really necessary to check the stdc version and/or C++ version? Does it cause a problem if we unconditionally define thread_local?

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h]: Is the forced alignment necessary? Remember that MSVC is not the only win32 compiler, so if we really need to force alignment, we'd have to also make it work for gcc/mingw at least. That could be a spearate commit later on, I could even add it myself at some point and test with a cross-compiler, but if it can be avoided altogether, it would be better.

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h]: Are these (`thrd_init`/`thrd_destroy`/`thrd_self_register`) in the standard? It's been a while since I've read it. If not, then isn't there a way to avoid them, like lazy initialization on first use?

If they can't be avoided, they should have distinct prefixes from the standard functions. Maybe `c11threads_` to make it obvious this is just an implementation detail of this library and not a standard function which can be used on portable programs. Also they should be hidden in the _win32.c file, if they're not meant to be called by users.

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h]: I don't think there's much point in adding that ifdef. What I meant by supporting older versions was that at some point in the future, it would be nice to either change the code to not use any vista APIs at all, or add detection and fallbacks for older win32. Until then it suffices to add a note to the README that the current win32 implementation requires vista or later.

File: c11threads_win32.c
Comment (jtsiomb) [inline c11threads_win32.c]: Put MSVC pragmas behind an ifdef. But in this case I'd just say drop them altogether. Users can tweak their warning levels to avoid annoying warnings.

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h]: There's no return for the non-win32 version. But is this function necessary at all? And if it's necessary only for the win32 code, maybe it should be moved into c11trheads_win32.c

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h:500]: TIME_UTC is only available in C11. I'd rather have this code be able to build with C99 or even C89 compilers if at all possible. Is this check necessary? What's the purpose of it?

**Review by jtsiomb — CHANGES_REQUESTED** [review]
Thank you. This is close to what I had in mind, but I'd like to discuss a couple of points before merging.

My main objection with the current version is the myrriad ifdefs required all over the place. I would prefer it if there was just a big:

    #if defined(_WIN32) && !defined(C11THREADS_PTHREAD_WIN32)
    /* just forward declarations for the functions in c11threads_win32.c */
    ...
    int thrd_exit(int res);
    ...
    #else
    /* thin wrappers for the POSIX case */
    ...
    static inline void thrd_exit(int res)
    {
        pthread_exit((void*)(intptr_t)res);
    }
    ...
    #endif

No need to have ifdefs in every function, and no need for extra leading-underscore `_thrd_*`  functions for the win32 versions. Unless I'm missing some complication you were trying to solve?

For the rest I'll add inline comments at the relevant parts of the changes.

File: c11threads.h
Comment (oold) [reply c11threads.h:500]: Clicked the wrong button... TL;DR of what I just wrote:

At least one base is needed. The function is supposed to return 0 on failure and `base` on success. Microsoft, for example, expects `TIME_UTC` and their function fails if anything else is provided as base.

Further documentation: https://en.cppreference.com/w/c/chrono/timespec_get

File: c11threads.h
Comment (oold) [reply c11threads.h]: Correct, I forgot to add the `return`. The function is in here to make it available by including the header. I could just have a declaration here, but felt it was cleaner to have a public function and an implementation.

File: c11threads.h
Comment (oold) [reply c11threads.h]: These aren't standard and I'll rename them. I was going with lazy initialization first, but when I noticed we'd have to free some resources anyway with some kind of dtor function figured I'd rather have the user initialize once and not bother with lazy initialization. I think that wouldn't be the right way because it could lead to delays. In a dynamic library, this would be done on load, and in the C library, it would be done on startup, so I decided to mirror that as best as possible.

File: c11threads_win32.c
Comment (oold) [reply c11threads_win32.c]: I'll be guarding them from non-Microsoft compilers.

File: c11threads.h
Comment (oold) [reply c11threads.h]: I see, I misunderstood. I still think it would be nice to provide the library without `cnd_...()` and `call_once()` for those on older Windows versions. From what I've gathered, `call_once()` can be replaced with some atomic functions and polling and I could easily implement that. But the condition variable functions are problematic. We simply don't have atomic wait-and-release operations prior to Vista.

File: c11threads.h
Comment (oold) [reply c11threads.h]: I had copied that from the declaration, but it really isn't necessary. I'll remove it.

File: c11threads.h
Comment (oold) [reply c11threads.h:27]: Defining the macro for C23 or newer would be non-standard, so I don't think that would be a good idea. Likewise, I believe defining a macro whose name is identical to a reserved keyword is not great.

**Comment by oold** [comment]
> My main objection with the current version is the myrriad ifdefs required all over the place.

I can consolidate the Win32-specific functions in the header. That seems like a good idea.

> No need to have ifdefs in every function, and no need for extra leading-underscore `_thrd_*` functions for the win32 versions. Unless I'm missing some complication you were trying to solve?

I've chosen the underscore to emphasize that those are internal functions, much like the C library using underscore and capital character or double underscore. I think that's better for debugging and navigating the code.

File: c11threads.h
Comment (jtsiomb) [reply c11threads.h]: You're not wrong. I suggest picking a less verbose name for it then like C11THREADS_OLD_WINAPI, or maybe something more directly related to what you're disabling, like C11THREADS_NO_COND / NO_ONCE + documentation about when to define them?

While there are no  condvar-equivalent before vista, I have implemented condvars on windows once upon a time, with a combination of a critical section, a semaphore, an event and SignalObjectAndWait... I could bundle up the code and attach it here if you're interested, but the comment at the top says: 

    /* POSIX thread condvar (pthread_cond_*) implementation based on:
     * "Strategies for Implementing POSIX Condition Variables on Win32"
     * by Douglas C. Schmidt and Irfan Pyarali.
     */

Edit: no need to change your code to use any of that hackery, I'm just saying there are other options for an eventual old-win32 fallback, if it can be detected somehow.

Edit2: [here's my old pthread implementation](https://github.com/jtsiomb/c11threads/files/10316770/pthread_win32.tar.gz)

File: c11threads.h
Comment (jtsiomb) [reply c11threads.h]: I'd rather not require the user to call any extra functions to use this, other than what they'd normally use if their standard library had native c11 threads support.

Is there a reason why we can't let those data structures be for the whole lifetime of the process and skip the destructor?

File: c11threads.h
Comment (jtsiomb) [reply c11threads.h:27]: I'm not fussed about that. We're not writing a specific C-standard libc after all. And we are defining all those things even if the system libc is C99 and would not have them anyway.

File: c11threads.h
Comment (jtsiomb) [inline c11threads.h]: Why are there a bunch of whitespace changes from this point on?

**Comment by jtsiomb** [comment]
> >  and no need for extra leading-underscore `_thrd_*` functions for the win32 versions. Unless I'm missing some complication you were trying to solve?
> 
> I've chosen the underscore to emphasize that those are internal functions, much like the C library using underscore and capital character or double underscore. I think that's better for debugging and navigating the code.

Right, but why have separate functions being called by one-line static inlines anyway? Just make the win32 implementation have the standard name to begin with.

File: c11threads.h
Comment (oold) [reply c11threads.h]: Visual Studio seems to have reformatted those lines. I'll remove the spaces in another commit.

**Comment by oold** [comment]
> Right, but why have separate functions being called by one-line static inlines anyway? Just make the win32 implementation have the standard name to begin with.

You're right. I'll drop the inline functions where appropriate.

File: c11threads.h
Comment (oold) [reply c11threads.h:27]: I'll define it always then unless it was already defined. Users can `#undef` it if they don't need it.

File: c11threads.h
Comment (oold) [reply c11threads.h]: A possible scenario could be that we're inside a library that needs to be unloaded.

File: c11threads.h
Comment (oold) [reply c11threads.h]: I'll go with `C11THREADS_NO_COND` for now and research the alternative implementation.

File: c11threads.h
Comment (oold) [reply c11threads.h:500]: Forgot to mention: c11threads.h will define `TIME_UTC` if it's not already defined.

File: c11threads.h
Comment (oold) [reply c11threads.h]: Those functions were moved. Solved.

File: c11threads.h
Comment (jtsiomb) [reply c11threads.h]: True, but that's a corner case. Let's keep the destroy function then (and the init for symmetry), but make init optional by lazy-initializing when necessary, and leave it to the user if they need to bother with destroy or not. But do not make these prominent. Put the win32 implementation-specific functions after the declarations of all the standard functions in the header file.

File: c11threads.h
Comment (jtsiomb) [reply c11threads.h:500]: That's perfectly fine then. Because previously it broke the build here, with TIME_UTC undefined, while compiling with `-std=gnu99` (C99 + extensions).

File: c11threads.h
Comment (oold) [reply c11threads.h]: Will do.

**Comment by jtsiomb** [comment]
I see you added a lot of commits. Can you summarize where we're at? Should I review the whole thing again and merge if everything is ok, or do you want to make more changes?

I see you also added a pre-vista condvar fallback. Nice! Did you use anything from my example or the article referenced by the comment? Or is it a different approach?

**Comment by oold** [comment]
I'm still thinking about the current implementation and will wait a few days before marking this as ready. Could be that I'll suddenly figure out a better way to implement some of this.

For the condvar implementation, I did copy some of your code first, but decided to make a bunch of changes, mostly due to this implementation using `CRITICAL_SECTION` as the mutex type. I've ended up with an approach that uses two events and a count variable. I hope we're not missing any events (shouldn't be) or end up busy-looping (my first approach looked kinda suspicious, but it should be fixed now). Still pondering whether to always reset the signal event, but I think the single spurious wakeup it could cause isn't a major issue.

**Comment by jtsiomb** [comment]
Sure, take your time.

I can't remember the tradeoffs involved in the decisions behind that code any more, but I do remember that the article compared multiple approaches, so it might still be worth a reading.

**Comment by oold** [comment]
I believe the branch is now in a reviewable state. There shouldn't be much left to be done. I've tested the code with VS 2022 (x64/x86), VS 2022 clang-cl (x64/x86), VS 2017 (x64/x86), and VS 2012 (x86). The worst offender was VS 2012. That C compiler is completely non-compliant and the C library doesn't even have a `struct timespec`. I've also cranked up the thread amount to 128 while testing my condvar implementation and there was nothing out of the ordinary to be observed.

**Comment by jtsiomb** [comment]
Excellent! Thank you for your effort. I'm sure win32 support will prove useful to many others.

There are some minor stylistic changed I'd like to do, mainly some simplifications to reduce the bulk of the header file, and some more functions which I think should be moved to the C file. But there's no point in going back and forth with them. It's much more efficient to merge this and clean up afterwards.

**Review by jtsiomb — APPROVED** [review]

**Comment by oold** [comment]
Thank you for the approval. It's probably best to squash the commits on the branch when merging.

**Comment by oold** [comment]
> some more functions which I think should be moved to the C file

The remaining functions can't really be moved, though. One could define `_USE_32BIT_TIME_T` per file, so the header has to redirect to the correct internal function for the specific local size of `tv_sec`.

**Comment by jtsiomb** [comment]
> The remaining functions can't really be moved, though. One could define `_USE_32BIT_TIME_T` per file, so the header has to redirect to the correct internal function for the specific local size of `tv_sec`.

Really? Per file? That sounds awful... I don't like it very much but I guess I'll leave them in the header file then.

**Comment by oold** [comment]
Microsoft's time.h header defines `struct timespec` and `timespec_get()` according to that definition. `timespec_get()` is `static inline`. Who would've thought?

**Comment by jtsiomb** [comment]
I squashed all your commits into one, and added a couple extra commits of my own on top to tweak things. Merged into master.

I mainly wanted to simplify the header file somewhat, having one huge ifdef instead of multiple ones, and remove the need for some of the macros like the *STATIC_INLINE one. I also tweaked the inline detection conditions slightly, leaving it at the end to the user to disable inline completely on the rare cases where a compiler does not support inline under any shape or form, or otherwise to define C11THREADS_INLINE to whatever is necessary.

I also tried to increase compatibility with older compilers. After some of my tweaks which, for the most part I think should not introduce any problems, I managed to compile this on VC2005 and MSVC6. For MSVC6 I had to conditionally side-step the register function with the `OpenThread` call, which apparently is only available on winXP and later APIs. I'm not sure what effect this will have, but at least it compiles and it appears to work well enough for the test to run. I'll probably have to revisit this particular hack at some point to see if it can be improved with a more appropriate fallback for older systems.

Thanks again for your contributions, and feel free to open further pull requests, or discuss this further.

**Pull request closed without merging**
